File: python/examples/sandbox_pty.py
"""Demonstrates the PTY (pseudo-terminal) namespace on a sandbox.

Creating an interactive session, streaming output, sending input, resizing,
and tearing it down. Mirrors ``js/examples/sandbox-pty.ts``.

Requires the ``websocket-client`` package (a regular dependency of
``lightning-sdk``).

Usage (from the ``python/`` directory)::

  LIGHTNING_SANDBOX_API_KEY=... python examples/sandbox_pty.py
"""

from __future__ import annotations

import sys
import time

from lightning_sdk.sandbox import (
    PtyConnectOpts,
    PtyCreateOpts,
    PtyHandle,
    SandboxInstance,
)


def main() -> None:
    sandbox = SandboxInstance.create(
        name=f"pty-example-{int(time.time())}",
        instance_type="cpu-2",
        cluster_id="baremetal",
    )
    print(f"Created sandbox: id={sandbox.sandbox_id} cluster={sandbox.cluster_id}")

    # --- Create an interactive PTY session -------------------------------
    print("\n--- Interactive PTY ---")

    pty = sandbox.process.create_pty(
        PtyCreateOpts(
            session_name="main-default",
            cluster_id=sandbox.cluster_id,
            cwd="/root",
            cols=120,
            rows=30,
            envs={"TERM": "xterm-256color"},
            # `on_data` is omitted: the SDK defaults to `write_to_stdout`,
            # which writes the raw shell bytes to `sys.stdout` and flushes
            # per chunk (mirrors Node's `process.stdout.write` on a TTY).
        )
    )

    # Surface handshake failures fast instead of hanging. Set
    # LIGHTNING_SDK_PTY_DEBUG=1 to see the wire trace if this raises.
    pty.wait_for_connection(timeout=30)
    pty.send_input("uname -a\n")
    pty.send_input("ls -la /\n")

    # Pretend the user just resized their terminal window.
    pty.resize(150, 40)

    # Cleanly exit the shell so .wait() returns.
    pty.send_input("exit\n")

    result = pty.wait()
    print(f"\nPTY closed: exit_code={result.exit_code} error={result.error}")

    # --- Listing / inspecting / killing sessions -------------------------
    print("\n--- Session bookkeeping ---")

    sessions = sandbox.process.list_pty_sessions()
    print(f"Active sessions: {len(sessions)}")
    for s in sessions:
        print(f"  {s.id} active={s.active} {s.cols}x{s.rows or '?'}")

    # Spawn a background loop and demonstrate kill_pty_session.
    # `PtyHandle.discard` opts out of the default stdout sink so the background
    # session's output doesn't interleave with the rest of this script.
    bg = sandbox.process.create_pty(
        PtyCreateOpts(
            session_name="main-background",
            cluster_id=sandbox.cluster_id,
            on_data=PtyHandle.discard,
        )
    )
    bg.wait_for_connection()
    bg.send_input("while true; do echo tick; sleep 1; done\n")
    time.sleep(2)
    sandbox.process.kill_pty_session("main-background")
    print("Killed background session")

    # `connect_pty` reuses an existing session within the same process; the
    # explicit reattach below works once cross-process persistence (`screen`
    # in the runtime image) is wired up.
    _ = PtyConnectOpts  # imported for the README example; suppress F401

    sandbox.delete()
    print("\nSandbox deleted")


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        print(err, file=sys.stderr)
        raise SystemExit(1) from err

File: js/examples/sandbox-pty.ts
/// <reference types="node" />
/**
 * Demonstrates the PTY (pseudo-terminal) namespace on a sandbox:
 * creating an interactive session, streaming output, sending input,
 * resizing, and tearing it down.
 *
 * Requires Node 22+ for the built-in WebSocket global.
 *
 * Usage:
 *   LIGHTNING_API_KEY=... npx tsx examples/sandbox-pty.ts
 */
import { Sandbox } from "../src/index.js";

async function main() {
  const sandbox = await Sandbox.create({
    name: "pty-example",
    instanceType: "cpu-2",
  });
  console.log("Created sandbox:", sandbox.sandboxId);

  // --- Create an interactive PTY session -----------------------------------
  console.log("\n--- Interactive PTY ---");

  const pty = await sandbox.process.createPty({
    sessionName: "main-default",
    cwd: "/root",
    cols: 120,
    rows: 30,
    envs: { TERM: "xterm-256color" },
    // `onData` is omitted: the SDK defaults to `writeToStdout`, which forwards
    // raw shell bytes to `process.stdout` (mirrors Python's default).
  });

  await pty.waitForConnection();

  await pty.sendInput("uname -a\n");
  await pty.sendInput("ls -la /\n");

  // Pretend the user just resized their terminal window.
  await pty.resize(150, 40);

  // Give the shell a moment to run the commands and stream their output back.
  await new Promise((r) => setTimeout(r, 2000));

  // A PTY is backed by a persistent `screen` session on the sandbox, so typing
  // `exit` only drops the *inner* shell -- the attach WebSocket stays open and
  // `pty.wait()` would block forever. Detach explicitly instead: `disconnect()`
  // closes the socket client-side, which is what makes `wait()` resolve. (To
  // also tear down the remote screen session, use `killPtySession()`, as the
  // background-session demo below does.)
  await pty.disconnect();

  const result = await pty.wait();
  console.log(`\nPTY closed: exitCode=${result.exitCode} error=${result.error}`);

  // --- Listing / inspecting / killing sessions -----------------------------
  console.log("\n--- Session bookkeeping ---");

  const sessions = await sandbox.process.listPtySessions();
  console.log(`Active sessions: ${sessions.length}`);
  for (const s of sessions) {
    console.log(`  ${s.id} active=${s.active} ${s.cols}x${s.rows ?? "?"}`);
  }

  // Spawn a background loop and demonstrate killPtySession.
  const bg = await sandbox.process.createPty({
    sessionName: "main-background",
    onData: () => { },
  });
  await bg.waitForConnection();
  await bg.sendInput("while true; do echo tick; sleep 1; done\n");
  await new Promise((r) => setTimeout(r, 2000));
  await sandbox.process.killPtySession("main-background");
  console.log("Killed background session");

  await sandbox.delete();
  console.log("\nSandbox deleted");
}

main().catch((err) => {
  console.error(err);
  process.exitCode = 1;
});
